{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "from sklearn.metrics.pairwise import cosine_similarity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "information_train = pd.read_csv('train/information_train.csv',delimiter='\\t')\n",
    "train = pd.read_csv('train/train.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>abstract</th>\n",
       "      <th>article_title</th>\n",
       "      <th>pmid</th>\n",
       "      <th>pub_date</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>We describe the Phase II HapMap, which charact...</td>\n",
       "      <td>A second generation human haplotype map of ove...</td>\n",
       "      <td>17943122</td>\n",
       "      <td>2009-05-21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Here we present the first diploid genome seque...</td>\n",
       "      <td>The diploid genome sequence of an Asian indivi...</td>\n",
       "      <td>18987735</td>\n",
       "      <td>2009-07-21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Following recent success in genome-wide associ...</td>\n",
       "      <td>Deletion polymorphism upstream of IRGM associa...</td>\n",
       "      <td>19165925</td>\n",
       "      <td>2009-07-30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Bowtie: a new ultrafast memory-efficient tool ...</td>\n",
       "      <td>Ultrafast and memory-efficient alignment of sh...</td>\n",
       "      <td>19261174</td>\n",
       "      <td>2009-12-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Motivation: A new protocol for sequencing the ...</td>\n",
       "      <td>TopHat: discovering splice junctions with RNA-Seq</td>\n",
       "      <td>19289445</td>\n",
       "      <td>2009-05-01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            abstract  \\\n",
       "0  We describe the Phase II HapMap, which charact...   \n",
       "1  Here we present the first diploid genome seque...   \n",
       "2  Following recent success in genome-wide associ...   \n",
       "3  Bowtie: a new ultrafast memory-efficient tool ...   \n",
       "4  Motivation: A new protocol for sequencing the ...   \n",
       "\n",
       "                                       article_title      pmid    pub_date  \n",
       "0  A second generation human haplotype map of ove...  17943122  2009-05-21  \n",
       "1  The diploid genome sequence of an Asian indivi...  18987735  2009-07-21  \n",
       "2  Deletion polymorphism upstream of IRGM associa...  19165925  2009-07-30  \n",
       "3  Ultrafast and memory-efficient alignment of sh...  19261174  2009-12-01  \n",
       "4  TopHat: discovering splice junctions with RNA-Seq  19289445  2009-05-01  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set_14_train = information_train[information_train.set==14]\n",
    "set_14_train.reset_index(drop=True, inplace=True)\n",
    "set_14_train = set_14_train.drop(['full_Text','author_str','set'], 1)\n",
    "set_14_train.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "tfidf_vectorizer_title = TfidfVectorizer(analyzer='word',ngram_range=(1,3),stop_words='english',norm='l2',max_features=50)\n",
    "X_title = tfidf_vectorizer_title.fit_transform(set_14_train['article_title'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "tfidf_vectorizer_abstract = TfidfVectorizer(analyzer='word',ngram_range=(1,2),stop_words='english',norm='l2',max_features=200)\n",
    "X = tfidf_vectorizer_abstract.fit_transform(set_14_train['abstract'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<1x50 sparse matrix of type '<class 'numpy.float64'>'\n",
       "\twith 1 stored elements in Compressed Sparse Row format>"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_title[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def merging_data_frame(particular_row_series,df_remaining_rows):\n",
    "    \n",
    "    duplicate_frequency =df_remaining_rows.shape[0] - 1\n",
    "    df_remaining_rows = df_remaining_rows.rename(columns={'abstract': 'abstract_1', 'article_title': 'article_title_1','pmid': 'pmid_1','pub_date': 'pub_date_1'})\n",
    "    df_particular_row = pd.DataFrame(particular_row_series).transpose()\n",
    "    df_particular_row = df_particular_row.append([df_particular_row]*duplicate_frequency,ignore_index=True)\n",
    "    \n",
    "\n",
    "    df_particular_row.reset_index(drop=True, inplace=True)\n",
    "    df_remaining_rows.reset_index(drop=True, inplace=True)\n",
    "\n",
    "    result = pd.concat([df_particular_row, df_remaining_rows], axis=1)\n",
    "    master_frames.append(result)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>abstract</th>\n",
       "      <th>article_title</th>\n",
       "      <th>pmid</th>\n",
       "      <th>pub_date</th>\n",
       "      <th>abstract_1</th>\n",
       "      <th>article_title_1</th>\n",
       "      <th>pmid_1</th>\n",
       "      <th>pub_date_1</th>\n",
       "      <th>differnce_days</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>We describe the Phase II HapMap, which charact...</td>\n",
       "      <td>A second generation human haplotype map of ove...</td>\n",
       "      <td>17943122</td>\n",
       "      <td>2009-05-21</td>\n",
       "      <td>Motivation: A new protocol for sequencing the ...</td>\n",
       "      <td>TopHat: discovering splice junctions with RNA-Seq</td>\n",
       "      <td>19289445</td>\n",
       "      <td>2009-05-01</td>\n",
       "      <td>20.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>We describe the Phase II HapMap, which charact...</td>\n",
       "      <td>A second generation human haplotype map of ove...</td>\n",
       "      <td>17943122</td>\n",
       "      <td>2009-05-21</td>\n",
       "      <td>The recycling of the amyloid precursor protein...</td>\n",
       "      <td>The neuronal sortilin-related receptor SORL1 i...</td>\n",
       "      <td>17220890</td>\n",
       "      <td>2009-01-16</td>\n",
       "      <td>125.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>We describe the Phase II HapMap, which charact...</td>\n",
       "      <td>A second generation human haplotype map of ove...</td>\n",
       "      <td>17943122</td>\n",
       "      <td>2009-05-21</td>\n",
       "      <td>There is increasing evidence that genome-wide ...</td>\n",
       "      <td>Genome-wide association study of 14,000 cases ...</td>\n",
       "      <td>17554300</td>\n",
       "      <td>2009-05-20</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>We describe the Phase II HapMap, which charact...</td>\n",
       "      <td>A second generation human haplotype map of ove...</td>\n",
       "      <td>17943122</td>\n",
       "      <td>2009-05-21</td>\n",
       "      <td>The Wellcome Trust Case Control Consortium (WT...</td>\n",
       "      <td>Robust associations of four new chromosome reg...</td>\n",
       "      <td>17554260</td>\n",
       "      <td>2007-06-21</td>\n",
       "      <td>700.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>We describe the Phase II HapMap, which charact...</td>\n",
       "      <td>A second generation human haplotype map of ove...</td>\n",
       "      <td>17943122</td>\n",
       "      <td>2009-05-21</td>\n",
       "      <td>TAR DNA-binding protein 43 (TDP-43) is a major...</td>\n",
       "      <td>TDP-43 in Familial and Sporadic Frontotemporal...</td>\n",
       "      <td>17591968</td>\n",
       "      <td>2007-07-01</td>\n",
       "      <td>690.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            abstract  \\\n",
       "0  We describe the Phase II HapMap, which charact...   \n",
       "1  We describe the Phase II HapMap, which charact...   \n",
       "2  We describe the Phase II HapMap, which charact...   \n",
       "3  We describe the Phase II HapMap, which charact...   \n",
       "4  We describe the Phase II HapMap, which charact...   \n",
       "\n",
       "                                       article_title      pmid   pub_date  \\\n",
       "0  A second generation human haplotype map of ove...  17943122 2009-05-21   \n",
       "1  A second generation human haplotype map of ove...  17943122 2009-05-21   \n",
       "2  A second generation human haplotype map of ove...  17943122 2009-05-21   \n",
       "3  A second generation human haplotype map of ove...  17943122 2009-05-21   \n",
       "4  A second generation human haplotype map of ove...  17943122 2009-05-21   \n",
       "\n",
       "                                          abstract_1  \\\n",
       "0  Motivation: A new protocol for sequencing the ...   \n",
       "1  The recycling of the amyloid precursor protein...   \n",
       "2  There is increasing evidence that genome-wide ...   \n",
       "3  The Wellcome Trust Case Control Consortium (WT...   \n",
       "4  TAR DNA-binding protein 43 (TDP-43) is a major...   \n",
       "\n",
       "                                     article_title_1    pmid_1 pub_date_1  \\\n",
       "0  TopHat: discovering splice junctions with RNA-Seq  19289445 2009-05-01   \n",
       "1  The neuronal sortilin-related receptor SORL1 i...  17220890 2009-01-16   \n",
       "2  Genome-wide association study of 14,000 cases ...  17554300 2009-05-20   \n",
       "3  Robust associations of four new chromosome reg...  17554260 2007-06-21   \n",
       "4  TDP-43 in Familial and Sporadic Frontotemporal...  17591968 2007-07-01   \n",
       "\n",
       "   differnce_days  \n",
       "0            20.0  \n",
       "1           125.0  \n",
       "2             1.0  \n",
       "3           700.0  \n",
       "4           690.0  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "master_frames = []\n",
    "for index,row in set_14_train.iterrows():\n",
    "    merging_data_frame(set_14_train.iloc[index],set_14_train.drop(set_14_train.index[[index]]))\n",
    "combined_df = pd.concat(master_frames,ignore_index=True)\n",
    "\n",
    "combined_df.pub_date = pd.to_datetime(combined_df.pub_date)\n",
    "combined_df.pub_date_1 = pd.to_datetime(combined_df.pub_date_1)\n",
    "\n",
    "combined_df['differnce_days'] = combined_df['pub_date'] - combined_df['pub_date_1']\n",
    "combined_df['differnce_days'] = combined_df['differnce_days'] / (np.timedelta64(1, 'D'))\n",
    "\n",
    "combined_df = combined_df[combined_df > 0].dropna()\n",
    "combined_df.reset_index(drop=True,inplace=True)\n",
    "\n",
    "combined_df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def cosine_similarity_abstract(row):\n",
    "    \n",
    "    index_1 = set_14_train.pmid[set_14_train.pmid == row['pmid']].index.tolist()[0]\n",
    "    index_2 = set_14_train.pmid[set_14_train.pmid == row['pmid_1']].index.tolist()[0]\n",
    "    similarity_score = cosine_similarity(X[index_1],X[index_2])\n",
    "    return similarity_score[0][0]\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def cosine_similarity_title(row):\n",
    "    \n",
    "    index_1 = set_14_train.pmid[set_14_train.pmid == row['pmid']].index.tolist()[0]\n",
    "    index_2 = set_14_train.pmid[set_14_train.pmid == row['pmid_1']].index.tolist()[0]\n",
    "    similarity_score = cosine_similarity(X_title[index_1],X_title[index_2])\n",
    "    return similarity_score[0][0]\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>abstract</th>\n",
       "      <th>article_title</th>\n",
       "      <th>pmid</th>\n",
       "      <th>pub_date</th>\n",
       "      <th>abstract_1</th>\n",
       "      <th>article_title_1</th>\n",
       "      <th>pmid_1</th>\n",
       "      <th>pub_date_1</th>\n",
       "      <th>differnce_days</th>\n",
       "      <th>similarity_title</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>We describe the Phase II HapMap, which charact...</td>\n",
       "      <td>A second generation human haplotype map of ove...</td>\n",
       "      <td>17943122</td>\n",
       "      <td>2009-05-21</td>\n",
       "      <td>Motivation: A new protocol for sequencing the ...</td>\n",
       "      <td>TopHat: discovering splice junctions with RNA-Seq</td>\n",
       "      <td>19289445</td>\n",
       "      <td>2009-05-01</td>\n",
       "      <td>20.0</td>\n",
       "      <td>0.152535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>We describe the Phase II HapMap, which charact...</td>\n",
       "      <td>A second generation human haplotype map of ove...</td>\n",
       "      <td>17943122</td>\n",
       "      <td>2009-05-21</td>\n",
       "      <td>The recycling of the amyloid precursor protein...</td>\n",
       "      <td>The neuronal sortilin-related receptor SORL1 i...</td>\n",
       "      <td>17220890</td>\n",
       "      <td>2009-01-16</td>\n",
       "      <td>125.0</td>\n",
       "      <td>0.072120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>We describe the Phase II HapMap, which charact...</td>\n",
       "      <td>A second generation human haplotype map of ove...</td>\n",
       "      <td>17943122</td>\n",
       "      <td>2009-05-21</td>\n",
       "      <td>There is increasing evidence that genome-wide ...</td>\n",
       "      <td>Genome-wide association study of 14,000 cases ...</td>\n",
       "      <td>17554300</td>\n",
       "      <td>2009-05-20</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.337733</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>We describe the Phase II HapMap, which charact...</td>\n",
       "      <td>A second generation human haplotype map of ove...</td>\n",
       "      <td>17943122</td>\n",
       "      <td>2009-05-21</td>\n",
       "      <td>The Wellcome Trust Case Control Consortium (WT...</td>\n",
       "      <td>Robust associations of four new chromosome reg...</td>\n",
       "      <td>17554260</td>\n",
       "      <td>2007-06-21</td>\n",
       "      <td>700.0</td>\n",
       "      <td>0.138157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>We describe the Phase II HapMap, which charact...</td>\n",
       "      <td>A second generation human haplotype map of ove...</td>\n",
       "      <td>17943122</td>\n",
       "      <td>2009-05-21</td>\n",
       "      <td>TAR DNA-binding protein 43 (TDP-43) is a major...</td>\n",
       "      <td>TDP-43 in Familial and Sporadic Frontotemporal...</td>\n",
       "      <td>17591968</td>\n",
       "      <td>2007-07-01</td>\n",
       "      <td>690.0</td>\n",
       "      <td>0.073892</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            abstract  \\\n",
       "0  We describe the Phase II HapMap, which charact...   \n",
       "1  We describe the Phase II HapMap, which charact...   \n",
       "2  We describe the Phase II HapMap, which charact...   \n",
       "3  We describe the Phase II HapMap, which charact...   \n",
       "4  We describe the Phase II HapMap, which charact...   \n",
       "\n",
       "                                       article_title      pmid   pub_date  \\\n",
       "0  A second generation human haplotype map of ove...  17943122 2009-05-21   \n",
       "1  A second generation human haplotype map of ove...  17943122 2009-05-21   \n",
       "2  A second generation human haplotype map of ove...  17943122 2009-05-21   \n",
       "3  A second generation human haplotype map of ove...  17943122 2009-05-21   \n",
       "4  A second generation human haplotype map of ove...  17943122 2009-05-21   \n",
       "\n",
       "                                          abstract_1  \\\n",
       "0  Motivation: A new protocol for sequencing the ...   \n",
       "1  The recycling of the amyloid precursor protein...   \n",
       "2  There is increasing evidence that genome-wide ...   \n",
       "3  The Wellcome Trust Case Control Consortium (WT...   \n",
       "4  TAR DNA-binding protein 43 (TDP-43) is a major...   \n",
       "\n",
       "                                     article_title_1    pmid_1 pub_date_1  \\\n",
       "0  TopHat: discovering splice junctions with RNA-Seq  19289445 2009-05-01   \n",
       "1  The neuronal sortilin-related receptor SORL1 i...  17220890 2009-01-16   \n",
       "2  Genome-wide association study of 14,000 cases ...  17554300 2009-05-20   \n",
       "3  Robust associations of four new chromosome reg...  17554260 2007-06-21   \n",
       "4  TDP-43 in Familial and Sporadic Frontotemporal...  17591968 2007-07-01   \n",
       "\n",
       "   differnce_days  similarity_title  \n",
       "0            20.0          0.152535  \n",
       "1           125.0          0.072120  \n",
       "2             1.0          0.337733  \n",
       "3           700.0          0.138157  \n",
       "4           690.0          0.073892  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df['similarity_title'] = combined_df.apply(cosine_similarity_abstract,axis=1)\n",
    "combined_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>abstract</th>\n",
       "      <th>article_title</th>\n",
       "      <th>pmid</th>\n",
       "      <th>pub_date</th>\n",
       "      <th>abstract_1</th>\n",
       "      <th>article_title_1</th>\n",
       "      <th>pmid_1</th>\n",
       "      <th>pub_date_1</th>\n",
       "      <th>differnce_days</th>\n",
       "      <th>similarity_abstract</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>We describe the Phase II HapMap, which charact...</td>\n",
       "      <td>A second generation human haplotype map of ove...</td>\n",
       "      <td>17943122</td>\n",
       "      <td>2009-05-21</td>\n",
       "      <td>Motivation: A new protocol for sequencing the ...</td>\n",
       "      <td>TopHat: discovering splice junctions with RNA-Seq</td>\n",
       "      <td>19289445</td>\n",
       "      <td>2009-05-01</td>\n",
       "      <td>20.0</td>\n",
       "      <td>0.015609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>We describe the Phase II HapMap, which charact...</td>\n",
       "      <td>A second generation human haplotype map of ove...</td>\n",
       "      <td>17943122</td>\n",
       "      <td>2009-05-21</td>\n",
       "      <td>The recycling of the amyloid precursor protein...</td>\n",
       "      <td>The neuronal sortilin-related receptor SORL1 i...</td>\n",
       "      <td>17220890</td>\n",
       "      <td>2009-01-16</td>\n",
       "      <td>125.0</td>\n",
       "      <td>0.011674</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>We describe the Phase II HapMap, which charact...</td>\n",
       "      <td>A second generation human haplotype map of ove...</td>\n",
       "      <td>17943122</td>\n",
       "      <td>2009-05-21</td>\n",
       "      <td>There is increasing evidence that genome-wide ...</td>\n",
       "      <td>Genome-wide association study of 14,000 cases ...</td>\n",
       "      <td>17554300</td>\n",
       "      <td>2009-05-20</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.049808</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>We describe the Phase II HapMap, which charact...</td>\n",
       "      <td>A second generation human haplotype map of ove...</td>\n",
       "      <td>17943122</td>\n",
       "      <td>2009-05-21</td>\n",
       "      <td>The Wellcome Trust Case Control Consortium (WT...</td>\n",
       "      <td>Robust associations of four new chromosome reg...</td>\n",
       "      <td>17554260</td>\n",
       "      <td>2007-06-21</td>\n",
       "      <td>700.0</td>\n",
       "      <td>0.034388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>We describe the Phase II HapMap, which charact...</td>\n",
       "      <td>A second generation human haplotype map of ove...</td>\n",
       "      <td>17943122</td>\n",
       "      <td>2009-05-21</td>\n",
       "      <td>TAR DNA-binding protein 43 (TDP-43) is a major...</td>\n",
       "      <td>TDP-43 in Familial and Sporadic Frontotemporal...</td>\n",
       "      <td>17591968</td>\n",
       "      <td>2007-07-01</td>\n",
       "      <td>690.0</td>\n",
       "      <td>0.011453</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            abstract  \\\n",
       "0  We describe the Phase II HapMap, which charact...   \n",
       "1  We describe the Phase II HapMap, which charact...   \n",
       "2  We describe the Phase II HapMap, which charact...   \n",
       "3  We describe the Phase II HapMap, which charact...   \n",
       "4  We describe the Phase II HapMap, which charact...   \n",
       "\n",
       "                                       article_title      pmid   pub_date  \\\n",
       "0  A second generation human haplotype map of ove...  17943122 2009-05-21   \n",
       "1  A second generation human haplotype map of ove...  17943122 2009-05-21   \n",
       "2  A second generation human haplotype map of ove...  17943122 2009-05-21   \n",
       "3  A second generation human haplotype map of ove...  17943122 2009-05-21   \n",
       "4  A second generation human haplotype map of ove...  17943122 2009-05-21   \n",
       "\n",
       "                                          abstract_1  \\\n",
       "0  Motivation: A new protocol for sequencing the ...   \n",
       "1  The recycling of the amyloid precursor protein...   \n",
       "2  There is increasing evidence that genome-wide ...   \n",
       "3  The Wellcome Trust Case Control Consortium (WT...   \n",
       "4  TAR DNA-binding protein 43 (TDP-43) is a major...   \n",
       "\n",
       "                                     article_title_1    pmid_1 pub_date_1  \\\n",
       "0  TopHat: discovering splice junctions with RNA-Seq  19289445 2009-05-01   \n",
       "1  The neuronal sortilin-related receptor SORL1 i...  17220890 2009-01-16   \n",
       "2  Genome-wide association study of 14,000 cases ...  17554300 2009-05-20   \n",
       "3  Robust associations of four new chromosome reg...  17554260 2007-06-21   \n",
       "4  TDP-43 in Familial and Sporadic Frontotemporal...  17591968 2007-07-01   \n",
       "\n",
       "   differnce_days  similarity_abstract  \n",
       "0            20.0             0.015609  \n",
       "1           125.0             0.011674  \n",
       "2             1.0             0.049808  \n",
       "3           700.0             0.034388  \n",
       "4           690.0             0.011453  "
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df['similarity_abstract'] = combined_df.apply(cosine_similarity_title,axis=1)\n",
    "combined_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def is_refernce(row):\n",
    "    index_train = train.pmid[train.pmid == row['pmid']].index.tolist()[0]\n",
    "    references_string = train.iloc[index_train]['ref_list']\n",
    "    references_string_clean = references_string.replace('[','').replace(']','').replace('\\'','').split(sep=',')\n",
    "    references_list = list(map(int, references_string_clean))\n",
    "    \n",
    "    if row['pmid_1'] in references_list:\n",
    "        return 1\n",
    "    return 0\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>abstract</th>\n",
       "      <th>article_title</th>\n",
       "      <th>pmid</th>\n",
       "      <th>pub_date</th>\n",
       "      <th>abstract_1</th>\n",
       "      <th>article_title_1</th>\n",
       "      <th>pmid_1</th>\n",
       "      <th>pub_date_1</th>\n",
       "      <th>differnce_days</th>\n",
       "      <th>similarity_abstract</th>\n",
       "      <th>is_reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>We describe the Phase II HapMap, which charact...</td>\n",
       "      <td>A second generation human haplotype map of ove...</td>\n",
       "      <td>17943122</td>\n",
       "      <td>2009-05-21</td>\n",
       "      <td>Motivation: A new protocol for sequencing the ...</td>\n",
       "      <td>TopHat: discovering splice junctions with RNA-Seq</td>\n",
       "      <td>19289445</td>\n",
       "      <td>2009-05-01</td>\n",
       "      <td>20.0</td>\n",
       "      <td>0.015609</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>We describe the Phase II HapMap, which charact...</td>\n",
       "      <td>A second generation human haplotype map of ove...</td>\n",
       "      <td>17943122</td>\n",
       "      <td>2009-05-21</td>\n",
       "      <td>The recycling of the amyloid precursor protein...</td>\n",
       "      <td>The neuronal sortilin-related receptor SORL1 i...</td>\n",
       "      <td>17220890</td>\n",
       "      <td>2009-01-16</td>\n",
       "      <td>125.0</td>\n",
       "      <td>0.011674</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>We describe the Phase II HapMap, which charact...</td>\n",
       "      <td>A second generation human haplotype map of ove...</td>\n",
       "      <td>17943122</td>\n",
       "      <td>2009-05-21</td>\n",
       "      <td>There is increasing evidence that genome-wide ...</td>\n",
       "      <td>Genome-wide association study of 14,000 cases ...</td>\n",
       "      <td>17554300</td>\n",
       "      <td>2009-05-20</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.049808</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>We describe the Phase II HapMap, which charact...</td>\n",
       "      <td>A second generation human haplotype map of ove...</td>\n",
       "      <td>17943122</td>\n",
       "      <td>2009-05-21</td>\n",
       "      <td>The Wellcome Trust Case Control Consortium (WT...</td>\n",
       "      <td>Robust associations of four new chromosome reg...</td>\n",
       "      <td>17554260</td>\n",
       "      <td>2007-06-21</td>\n",
       "      <td>700.0</td>\n",
       "      <td>0.034388</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>We describe the Phase II HapMap, which charact...</td>\n",
       "      <td>A second generation human haplotype map of ove...</td>\n",
       "      <td>17943122</td>\n",
       "      <td>2009-05-21</td>\n",
       "      <td>TAR DNA-binding protein 43 (TDP-43) is a major...</td>\n",
       "      <td>TDP-43 in Familial and Sporadic Frontotemporal...</td>\n",
       "      <td>17591968</td>\n",
       "      <td>2007-07-01</td>\n",
       "      <td>690.0</td>\n",
       "      <td>0.011453</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            abstract  \\\n",
       "0  We describe the Phase II HapMap, which charact...   \n",
       "1  We describe the Phase II HapMap, which charact...   \n",
       "2  We describe the Phase II HapMap, which charact...   \n",
       "3  We describe the Phase II HapMap, which charact...   \n",
       "4  We describe the Phase II HapMap, which charact...   \n",
       "\n",
       "                                       article_title      pmid   pub_date  \\\n",
       "0  A second generation human haplotype map of ove...  17943122 2009-05-21   \n",
       "1  A second generation human haplotype map of ove...  17943122 2009-05-21   \n",
       "2  A second generation human haplotype map of ove...  17943122 2009-05-21   \n",
       "3  A second generation human haplotype map of ove...  17943122 2009-05-21   \n",
       "4  A second generation human haplotype map of ove...  17943122 2009-05-21   \n",
       "\n",
       "                                          abstract_1  \\\n",
       "0  Motivation: A new protocol for sequencing the ...   \n",
       "1  The recycling of the amyloid precursor protein...   \n",
       "2  There is increasing evidence that genome-wide ...   \n",
       "3  The Wellcome Trust Case Control Consortium (WT...   \n",
       "4  TAR DNA-binding protein 43 (TDP-43) is a major...   \n",
       "\n",
       "                                     article_title_1    pmid_1 pub_date_1  \\\n",
       "0  TopHat: discovering splice junctions with RNA-Seq  19289445 2009-05-01   \n",
       "1  The neuronal sortilin-related receptor SORL1 i...  17220890 2009-01-16   \n",
       "2  Genome-wide association study of 14,000 cases ...  17554300 2009-05-20   \n",
       "3  Robust associations of four new chromosome reg...  17554260 2007-06-21   \n",
       "4  TDP-43 in Familial and Sporadic Frontotemporal...  17591968 2007-07-01   \n",
       "\n",
       "   differnce_days  similarity_abstract  is_reference  \n",
       "0            20.0             0.015609             0  \n",
       "1           125.0             0.011674             0  \n",
       "2             1.0             0.049808             1  \n",
       "3           700.0             0.034388             0  \n",
       "4           690.0             0.011453             0  "
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df['is_reference'] = combined_df.apply(is_refernce,axis=1)\n",
    "combined_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pmid</th>\n",
       "      <th>pmid_1</th>\n",
       "      <th>differnce_days</th>\n",
       "      <th>similarity_abstract</th>\n",
       "      <th>similarity_title</th>\n",
       "      <th>is_reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>17943122</td>\n",
       "      <td>19289445</td>\n",
       "      <td>20.0</td>\n",
       "      <td>0.015609</td>\n",
       "      <td>0.015609</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>17943122</td>\n",
       "      <td>17220890</td>\n",
       "      <td>125.0</td>\n",
       "      <td>0.011674</td>\n",
       "      <td>0.011674</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>17943122</td>\n",
       "      <td>17554300</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.049808</td>\n",
       "      <td>0.049808</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>17943122</td>\n",
       "      <td>17554260</td>\n",
       "      <td>700.0</td>\n",
       "      <td>0.034388</td>\n",
       "      <td>0.034388</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>17943122</td>\n",
       "      <td>17591968</td>\n",
       "      <td>690.0</td>\n",
       "      <td>0.011453</td>\n",
       "      <td>0.011453</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       pmid    pmid_1  differnce_days  similarity_abstract  similarity_title  \\\n",
       "0  17943122  19289445            20.0             0.015609          0.015609   \n",
       "1  17943122  17220890           125.0             0.011674          0.011674   \n",
       "2  17943122  17554300             1.0             0.049808          0.049808   \n",
       "3  17943122  17554260           700.0             0.034388          0.034388   \n",
       "4  17943122  17591968           690.0             0.011453          0.011453   \n",
       "\n",
       "   is_reference  \n",
       "0             0  \n",
       "1             0  \n",
       "2             1  \n",
       "3             0  \n",
       "4             0  "
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df_for_model = combined_df[['pmid','pmid_1','differnce_days','similarity_abstract','similarity_title','is_reference']]\n",
    "combined_df_for_model.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "predictors = combined_df_for_model.select_dtypes(['float64', 'int64', 'uint8']).columns.drop('is_reference')\n",
    "\n",
    "import xgboost as xgb\n",
    "\n",
    "y_train = combined_df_for_model.is_reference\n",
    "\n",
    "xgb_params = {\n",
    "   'eta': 0.3,\n",
    "   'max_depth': 5,\n",
    "   'subsample': 0.6,\n",
    "   'colsample_bytree': 1,\n",
    "   'lambda': 3,\n",
    "   'objective': 'binary:logistic',\n",
    "   'eval_metric': 'auc',\n",
    "   'silent': 0\n",
    "}\n",
    "\n",
    "dtrain = xgb.DMatrix(data=combined_df_for_model[predictors], label= y_train)\n",
    "\n",
    "\n",
    "num_rounds = 10000\n",
    "\n",
    "model_cv = xgb.cv(xgb_params, dtrain, num_rounds, nfold=10, early_stopping_rounds=20, verbose_eval=1)\n",
    "model = xgb.train(xgb_params, dtrain, num_boost_round = 44)\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "\n",
    "fig, ax = plt.subplots(1, 1, figsize=(16, 16))\n",
    "xgb.plot_importance(model, height=0.5, ax=ax)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "predictors = combined_df.select_dtypes(['float64', 'int64', 'uint8']).columns.drop('is_reference').drop('pmid_1')\n",
    "\n",
    "import xgboost as xgb\n",
    "\n",
    "y_train = combined_df.is_reference\n",
    "\n",
    "xgb_params = {\n",
    "   'eta': 0.3,\n",
    "   'max_depth': 5,\n",
    "   'subsample': 0.6,\n",
    "   'colsample_bytree': 1,\n",
    "   'lambda': 3,\n",
    "   'objective': 'binary:logistic',\n",
    "   'eval_metric': 'auc',\n",
    "   'silent': 0\n",
    "}\n",
    "\n",
    "dtrain = xgb.DMatrix(data=combined_df[predictors], label= y_train)\n",
    "\n",
    "\n",
    "num_rounds = 10000\n",
    "\n",
    "model_cv = xgb.cv(xgb_params, dtrain, num_rounds, nfold=10, early_stopping_rounds=20, verbose_eval=1)\n",
    "model = xgb.train(xgb_params, dtrain, num_boost_round = 44)\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "\n",
    "fig, ax = plt.subplots(1, 1, figsize=(16, 16))\n",
    "xgb.plot_importance(model, height=0.5, ax=ax)\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
